File: MP7/JavaTemplate/TablePartF.java
import java.io.IOException;

import org.apache.hadoop.conf.Configuration;

import org.apache.hadoop.hbase.Cell;
import org.apache.hadoop.hbase.CellUtil;
import org.apache.hadoop.hbase.HBaseConfiguration;
import org.apache.hadoop.hbase.HColumnDescriptor;
import org.apache.hadoop.hbase.HTableDescriptor;
import org.apache.hadoop.hbase.TableName;

import org.apache.hadoop.hbase.client.Admin;
import org.apache.hadoop.hbase.client.Connection;
import org.apache.hadoop.hbase.client.ConnectionFactory;
import org.apache.hadoop.hbase.client.Get;
import org.apache.hadoop.hbase.client.Put;
import org.apache.hadoop.hbase.client.Result;
import org.apache.hadoop.hbase.client.ResultScanner;
import org.apache.hadoop.hbase.client.Scan;
import org.apache.hadoop.hbase.client.Table;
import org.apache.hadoop.hbase.client.TableDescriptor;

import org.apache.hadoop.hbase.util.Bytes;

public class TablePartF{

   public static void main(String[] args) throws IOException {

	// TODO      
	// DON' CHANGE THE 'System.out.println(xxx)' OUTPUT PART
	// OR YOU WON'T RECEIVE POINTS FROM THE GRADER
	   
	//1. update the data in a particular cell using the put() method
	//2. Reading values from Result class object to retrieve all versions of all columns, with the most recent version coming first

	for (Cell cell : ???) {
		byte[] family = = ???;
		byte[] qualifier = = ???;
		byte[] value = = ???;
		long timestamp = = ???;

		System.out.println(
			"row: " + ??? +
			", family: " + ??? +
			", qualifier: " + ??? +
			", value: " + ??? +
			", timestamp: " + timestamp
		);
	}

   }
}

File: MP7/PythonTemplate/TablePartF.py
import happybase as hb

# DON'T CHANGE THE PRINT FORMAT, WHICH IS THE OUTPUT
# OR YOU WON'T RECEIVE POINTS FROM THE GRADER

# 1. update the data in a particular cell using the put() method
# 2. retrieve all versions of all columns, with the most recent version coming first

connection = hb.Connection('localhost')
connection.open()

table = connection.table('powers')

# Row key for the row to update
row_key = 'row7'

# Update the data in a particular cell
table.put(row_key, {'custom:color': 'purple'})

row = table.row(row_key, include_timestamp=True)
for column_family, data in row.items():
    #print(f"{column_family} ::: {data}")

    value, timestamp = data
    if column_family == b'custom:color':
        cells = table.cells(row=row_key, column='custom:color', versions=2, include_timestamp=True)
        #print(f"cells {cells}")
        for color, timestamp_inner in cells:
            ##print(f"color {color}")
            print("row: {}, column family:qualifier: {}, value: {}, timestamp: {}".format(row_key.encode('ascii'), column_family, color, timestamp_inner))
    else:
        print("row: {}, column family:qualifier: {}, value: {}, timestamp: {}".format(row_key.encode('ascii'), column_family, value, timestamp))
